api/src/main/java/org/pmiops/workbench/db/dao/NotebookCohortCacheServiceImpl.java

```diff
@@ -0,0 +1,124 @@
+      Timestamp lastAccessDateTime) {
+    WorkspaceUserRole workspaceUserRole =
+        getworkspaceUserRoleDao().findWorkspaceUserRolesByWorkspaceAndUser(workspace, user);
+    if (workspaceUserRole == null) {
+      throw new BadRequestException(String.format(
+          "No workspace %s was found for user %s ", workspace.getName(), user.getUserId()));
+    }
+    NotebookCohortCache cache = getDao().findByUserWorkspaceIdAndNotebookName
+        (workspaceUserRole, notebookName);
+    if (cache == null) {
+      handleWorkspaceUserLimitInCache(workspaceUserRole);
+      cache = new NotebookCohortCache();
+      cache.setUserWorkspaceId(workspaceUserRole);
+      cache.setCohortId(null);
+      cache.setNotebookName(notebookName);
+    }
+    cache.setLastAccessTime(lastAccessDateTime);
+    getDao().save(cache);
+  }
+
+  /**
+   * Checks if cohort is already in cache for given workspace and user
+   * if yes, update the lastmodified time only
+   * If no, check the number of cache entries for given workspace and user
+   * if it is above 10, delete the row with least lastAccessTime and add a new entry
+   * @param workspace
+   * @param user
+   * @param cohortId
+   * @param lastAccessDateTime
+   */
+  @Override
+  public void updateCohort(Workspace workspace, User user, long cohortId,
+      Timestamp lastAccessDateTime) {
+    WorkspaceUserRole workspaceUserRole =
+        getworkspaceUserRoleDao().findWorkspaceUserRolesByWorkspaceAndUser(workspace, user);
+    if (workspaceUserRole == null) {
+      throw new BadRequestException(String.format(
+          "No workspace %s was found for user %s ", workspace.getName(), user.getUserId()));
+    }
+    NotebookCohortCache cache = getDao().findByUserWorkspaceIdAndCohortId(workspaceUserRole, cohortId);
+    if (cache == null) {
+      handleWorkspaceUserLimitInCache(workspaceUserRole);
+      cache = new NotebookCohortCache();
+      cache.setUserWorkspaceId(workspaceUserRole);
+      cache.setCohortId(cohortId);
+      cache.setNotebookName(null);
+    }
+    cache.setLastAccessTime(lastAccessDateTime);
+    getDao().save(cache);
+  }
+
+  @Override
+  public void deleteNotebook(Workspace workspace, User user, String notebookName) {
+    WorkspaceUserRole workspaceUserRole =
+        getworkspaceUserRoleDao().findWorkspaceUserRolesByWorkspaceAndUser(workspace, user);
+    if (workspaceUserRole == null) {
+      throw new BadRequestException(String.format(
+          "User %s with workspace %s not found.", workspace.getName(), user.getUserId()));
+    }
+    getDao().
+        deleteNotebookCohortCacheByUserWorkspaceIdAndNotebookName(workspaceUserRole, notebookName);
+  }
+
+  //Delete the last access row for given workspace and user
+  private void handleWorkspaceUserLimitInCache(WorkspaceUserRole user_workspace) {
+    long count = getDao().countNotebookCohortCachesByUserWorkspaceId(user_workspace);
+    if (count == configProvider.get().notebookCohortCacheConfig.count) {
```

Comment:
Please use >=.

Optionally, wrap this whole thing in a `while` to handle going over the cache. This could matter if we changed the config value from 20 -> 10, for example.